cmd/snap/cmd_run.go

```diff
@@ -95,6 +97,38 @@ func (x *cmdRun) Execute(args []string) error {
 		x.Command = "shell"
 	}
 
+	// Check if snapd is starting up. If so wait because it means
+	// snapd is still re-generating security profiles. The
+	// snapd.service is configured so that it will enter a
+	// permanent failed state if it cannot be activated and at this
+	// point the code will continue (with potentially stale profiles
+	// but at least services will run).
+	for i := 0; i < 500; i++ {
+		cmd := exec.Command("systemctl", "show", "-pActiveState", "snapd.service")
+		raw, err := cmd.CombinedOutput()
+		output := strings.TrimSpace(string(raw))
+		// XXX: only needed because systemd 231 behave
+		//      different from 204,229,232. With those we get
+		//      the expected "ActiveState=inactive".
+		if err != nil && output == "Unit snapd.service could not be found." {
+			break
+		}
+		// XXX2: the backported systemd on 14.04 does not support
+		//       the dbus interface of systemd, only the private
+		//       socket which is only available as root, ignore
+		//       this error.
```

Comment:
In addition to not being supported on 14.04, it occurred to me that performing a DBus call in snap run is overhead we could avoid with a stamp file. I'm not sure if this fits all the properties you are looking for, but if instead of this DBus call snap run checks for the existence of a file (eg `/run/snapd/something-appropriate`) and waits on that, this would be a cheap check that would also work on 14.04. This stamp file would be created after profile generation finished and it could be removed on 'stop' so that 'snap run' during a core snap refresh would cause snap run to wait.